File: bin/translate_posts.rb
#!/usr/bin/env ruby

# Translate French posts to other languages using DeepL.
#
# Usage:
#   DEEPL_AUTH_KEY=<key> ruby bin/translate_posts.rb
#   ruby bin/translate_posts.rb               # reads key from .deepl_key if present
#   ruby bin/translate_posts.rb --dry-run     # list files that would be translated
#   ruby bin/translate_posts.rb               # read key from .deepl_key if present
#
# The script translates only the markdown body (everything after the front matter)
# and leaves the YAML front matter keys intact (except it updates the `title`).

require 'yaml'
require 'net/http'
require 'uri'
require 'json'
require 'optparse'
require 'time'
require 'fileutils'

options = {
  dry_run: false,
  langs: %w[en es],
  source_lang: 'FR',
}

OptionParser.new do |opts|
  opts.banner = "Usage: #{$PROGRAM_NAME} [options]"

  opts.on('-n', '--dry-run', 'Show which files would be translated without calling DeepL or writing output') do
    options[:dry_run] = true
  end

  opts.on('-l', '--langs LANGS', 'Comma-separated list of target languages (default: en,es)') do |v|
    options[:langs] = v.split(',').map(&:strip).reject(&:empty?)
  end

  opts.on('-h', '--help', 'Show help') do
    puts opts
    exit 0
  end
end.parse!

root = File.expand_path('..', __dir__)

# Prefer environment variable, but allow a local key file for safety (not checked in).
api_key = ENV['DEEPL_AUTH_KEY'] || ENV['DEEPL_API_KEY']
key_path = File.join(root, '.deepl_key')
if api_key.to_s.strip.empty? && File.exist?(key_path)
  api_key = File.read(key_path).strip
end

if api_key.to_s.strip.empty? && !options[:dry_run]
  warn "ERROR: Set DEEPL_AUTH_KEY (or DEEPL_API_KEY) in the environment, or create #{key_path} with your key."
  exit 1
end

api_url = ENV['DEEPL_API_URL'] || 'https://api-free.deepl.com/v2/translate'

# Helper: split YAML front matter + body
def split_front_matter(content)
  return [nil, content] unless content.start_with?("---\n")

  parts = content.split(/^---\s*\n/, 3)
  # parts: ["", "...yaml...", "body..."]
  if parts.size == 3
    [parts[1], parts[2]]
  else
    [nil, content]
  end
end

## Extract a contiguous grades block (lines with labels and numeric values) and return [before, grades_lines_array, after]
def extract_grades_block(text)
  lines = text.lines
  n = lines.length
  start_idx = nil
  end_idx = nil

  # be permissive: treat lines containing ": <number>" as grade lines
  grade_line_re = /:\s*[0-9]+(?:\.[0-9]+)?/

  (0...n).each do |i|
    if lines[i] =~ grade_line_re
      start_idx ||= i
      end_idx = i
    elsif start_idx && end_idx && i == end_idx + 1
      break
    end
  end

  return [text, [], ""] unless start_idx

  heading_idx = start_idx - 1
  if heading_idx >= 0 && lines[heading_idx] =~ /^\s*#+\s+/ && (start_idx - heading_idx) <= 2
    start_idx = heading_idx
  end

  before = lines[0...start_idx].join
  grades = lines[start_idx..end_idx].map(&:chomp)
  after = lines[(end_idx+1)..-1]&.join || ""
  [before, grades, after]
end

## Try to find a template labels list in existing translations for `lang`.
## Returns array of label strings (including formatting like underscores) or nil.
def find_template_labels(root, lang)
  dest_dir = File.join(root, '_i18n', lang, '_posts')
  return nil unless Dir.exist?(dest_dir)

  Dir.glob(File.join(dest_dir, '*.md')).each do |p|
    content = File.read(p)
    _fm, body = split_front_matter(content)
    before, grades, after = extract_grades_block(body)
    next if grades.empty?
    labels = grades.map do |line|
      parts = line.split(':', 2)
      parts[0].strip
    end
    return labels unless labels.empty?
  end
  nil
end

# DeepL translate helper
def deepl_translate(text, target_lang, source_lang, api_key, api_url, options)
  puts "Translating to #{target_lang} (#{text.length} chars)..." if options[:dry_run] == false
  uri = URI(api_url)
  req = Net::HTTP::Post.new(uri)
  req['Authorization'] = "DeepL-Auth-Key #{api_key}"
  req.set_form_data(
    'text' => text,
    'source_lang' => source_lang,
    'target_lang' => target_lang.upcase,
    'preserve_formatting' => '1',
    'split_sentences' => 'nonewlines'
  )

  res = Net::HTTP.start(
    uri.hostname,
    uri.port,
    use_ssl: uri.scheme == 'https',
    open_timeout: 10,
    read_timeout: 30
  ) do |http|
    http.request(req)
  end

  unless res.is_a?(Net::HTTPSuccess)
    raise "DeepL API error #{res.code}: #{res.body}"
  end

  body = JSON.parse(res.body)
  (body['translations'] || []).map { |t| t['text'] }.join("\n\n")
end

source_dir = File.join(root, '_i18n', 'fr', '_posts')

# Try to find a template labels list in existing translations for `lang`.
# Returns array of label strings (including formatting like underscores) or nil.

unless Dir.exist?(source_dir)
  warn "ERROR: source directory not found: #{source_dir}"
  exit 1
end

translated = []
skipped = []

Dir.glob(File.join(source_dir, '*.md')).sort.each do |src_path|
  filename = File.basename(src_path)
  content = File.read(src_path)

  front_matter_text, body = split_front_matter(content)
  front_matter = front_matter_text ? YAML.safe_load(front_matter_text) : {}

  options[:langs].each do |lang|

# Build a grades block using template labels (if provided) and the numeric values from source_grades_lines
def build_grades_block(template_labels, source_grades_lines)
  # If no template labels provided, we'll fall back to language defaults elsewhere.
  return source_grades_lines.join("\n") if source_grades_lines.empty?

  # Determine heading, labels and final label based on template_labels or language hints.
  # If template_labels is a Hash mapping (lang => labels), handle accordingly; otherwise
  # assume it's an Array of five labels.

  # Normalize input: if first line is a heading like '###', drop it but remember to emit localized heading.
  heading = nil
  lines = source_grades_lines.dup
  if lines[0] =~ /^\s*#+\s*/
    heading = lines.shift.strip
  end

  # Extract numeric values (in order) and final score/fraction
  numbers = []
  final_fraction = nil
  lines.each do |ln|
    # final line often contains bold markers and a fraction like 23/25
    if ln =~ /\*\*/
      m = ln.match(/([0-9]+(?:\.[0-9]+)?\s*\/\s*25)/)
      final_fraction = m ? m[1].gsub('\s','') : nil
      next
    end
    m = ln.match(/([0-9]+(?:\.[0-9]+)?)/)
    numbers << (m ? m[1] : '')
  end

  # If template_labels is an Array, use as-is; otherwise caller may supply nil.
  labels = template_labels.is_a?(Array) ? template_labels : nil

  # Attempt to infer language from labels if possible (look for obvious words)
  lang = nil
  if labels
    # quick heuristic
    first = labels[0].to_s.downcase
    lang = 'es' if first.include?('original') || first.include?('originalidad')
    lang = 'en' if first.include?('uniqu') || first.include?('uniqueness')
  end

  # Fallback label sets per language
  fallbacks = {
    'en' => ['Uniqueness', 'Finesse', 'Comfort', 'Intensity', 'Overall impression'],
    'es' => ['Originalidad', 'Fineza', 'Reconfortante', 'Intensidad', 'Impresión general']
  }

  # Choose labels: prefer provided template_labels, else fallbacks based on lang, else English.
  chosen = labels || (lang && fallbacks[lang]) || fallbacks['en']

  # Choose heading and final label by lang
  heading_map = { 'en' => '### Evaluation', 'es' => '### Evaluación' }
  final_map = { 'en' => '**Final evaluation**', 'es' => '**Nota final**' }
  heading_emit = heading_map[lang] || heading_map['en']
  final_label_emit = final_map[lang] || final_map['en']

  out = []
  out << heading_emit
  # Emit five labeled lines with single underscores and two trailing spaces to preserve linebreaks
  chosen.each_with_index do |lbl, idx|
    # sanitize label: remove surrounding underscores/asterisks and trailing colons
    clean = lbl.to_s.gsub(/^\s+|\s+$/, '').gsub(/^_+|_+$|^\*+|\*+$/, '').sub(/:$/, '')
    val = numbers[idx] || ''
    out << "_#{clean}_: #{val}  "
  end
  # Emit final fraction if found, else attempt to sum numbers
  if final_fraction
    out << "\n#{final_label_emit}: #{final_fraction}"
  else
    # compute a total if numbers look numeric
    if numbers.all? { |n| n.to_s =~ /^\d+(?:\.\d+)?$/ }
      total = numbers.map(&:to_f).sum
      total_s = (total % 1.0 == 0) ? total.to_i.to_s : total.to_s
      out << "\n#{final_label_emit}: #{total_s}/25"
    else
      out << "\n#{final_label_emit}:"
    end
  end

  out.join("\n")
end
    dest_dir = File.join(root, '_i18n', lang, '_posts')
    FileUtils.mkdir_p(dest_dir)
    dest_path = File.join(dest_dir, filename)

    if File.exist?(dest_path)
      skipped << dest_path
      next
    end

    if options[:dry_run]
      translated << dest_path
      next
    end

    # Extract grades block from source and avoid translating it
    before_text, source_grades_lines, after_text = extract_grades_block(body)

    # Translate body parts separately so we can reinsert a controlled grades block
    translated_before = before_text.strip.empty? ? "" : deepl_translate(before_text, lang, options[:source_lang], api_key, api_url, options)
    translated_after  = after_text.strip.empty?  ? "" : deepl_translate(after_text,  lang, options[:source_lang], api_key, api_url, options)

    # Build grades block from template labels if available (keeps numbers from source)
    template_labels = find_template_labels(root, lang)
    if source_grades_lines.empty?
      grades_block = ""
    else
      # build_grades_block defined below
      grades_block = build_grades_block(template_labels, source_grades_lines)
    end

    translated_body = [translated_before.to_s.strip, grades_block.to_s.strip, translated_after.to_s.strip].reject(&:empty?).join("\n\n")

    # Translate title if present
    translated_title = nil
    if front_matter.is_a?(Hash) && front_matter['title'].is_a?(String)
      translated_title = deepl_translate(front_matter['title'], lang, options[:source_lang], api_key, api_url, options).strip
      front_matter['title'] = translated_title
    end

    # Translate tags if present and ensure tags are an array
    if front_matter.is_a?(Hash) && front_matter['tags']
      tags = front_matter['tags'].is_a?(Array) ? front_matter['tags'] : [front_matter['tags'].to_s]
      translated_tags = tags.map do |t|
        t = t.to_s.strip
        next t if t.empty?

        # Hard-coded overrides for specific tags (avoid calling DeepL for these)
        lc = t.downcase
        if lc == 'noir'
          case lang
          when 'es'
            'Oscuro'
          when 'en'
            'Dark'
          else
            t
          end
        else
          deepl_translate(t, lang, options[:source_lang], api_key, api_url, options).strip
        end
      end
      front_matter['tags'] = translated_tags
      translated_tags_var = translated_tags
    end

    # Do not add translation metadata fields (translated_from/translated_at/translated_to)
    # — these were removed to keep generated front matter minimal.

    output = "---\n"
    # Always write the title as a quoted YAML scalar to match site conventions.
    # Use the translated_title (if set) as a reliable source; fall back to front_matter['title'].
    title_to_emit = translated_title || (front_matter.is_a?(Hash) && front_matter['title'])
    if title_to_emit
      title_val = title_to_emit.to_s
      escaped = title_val.gsub('"', '\\"')
      output += "title: \"#{escaped}\"\n"
      front_matter.delete('title') if front_matter.is_a?(Hash)
    end

    # Emit tags and categories as inline bracket lists to match site conventions,
    # then dump any remaining front_matter. We remove emitted keys to avoid duplication.
    if front_matter.is_a?(Hash)
      # Emit tags: handle both String and Array (use translated_tags_var if present)
      tags_val = defined?(translated_tags_var) ? translated_tags_var : front_matter['tags']
      if tags_val
        tags_arr = tags_val.is_a?(Array) ? tags_val : [tags_val.to_s]
        tags_arr = tags_arr.map(&:to_s).reject(&:empty?)
        unless tags_arr.empty?
          formatted = tags_arr.map { |t| "\"#{t.gsub('"', '\\\"')}\"" }.join(', ')
          output += "tags: [#{formatted}]\n"
        end
        front_matter.delete('tags')
      end

      # Emit categories: handle both String and Array
      cats_val = front_matter['categories']
      if cats_val
        cats_arr = cats_val.is_a?(Array) ? cats_val : [cats_val.to_s]
        cats_arr = cats_arr.map(&:to_s).reject(&:empty?)
        unless cats_arr.empty?
          formatted = cats_arr.map { |c| "\"#{c.gsub('"', '\\\"')}\"" }.join(', ')
          output += "categories: [#{formatted}]\n"
        end
        front_matter.delete('categories')
      end

      unless front_matter.empty?
        output += YAML.dump(front_matter).sub(/^---\n/, '')
      end
    end

    output += "---\n\n"
    output += translated_body.strip + "\n"

    File.write(dest_path, output)
    translated << dest_path
  end
end

puts "Translated #{translated.size} files (#{options[:langs].join(', ')})"
translated.each do |path|
  puts "  - #{path}"
end
puts "Skipped #{skipped.size} existing files" if skipped.any?

File: bin/translate_posts.py
#!/usr/bin/env python3
"""
Translate French posts to other languages using DeepL (Python port).

Usage:
  DEEPL_AUTH_KEY=<key> python3 bin/translate_posts.py
  python3 bin/translate_posts.py               # reads key from .deepl_key if present
  python3 bin/translate_posts.py --dry-run     # list files that would be translated
  python3 bin/translate_posts.py --langs en,es

Notes:
- This script intentionally does not overwrite existing translated files.
- It preserves numeric grades and applies localized labels.
"""

import argparse
import os
import sys
import json
import time
import re
from urllib import request, parse, error
from pathlib import Path

try:
    import yaml
    HAS_YAML = True
except Exception:
    HAS_YAML = False

DEFAULT_LANGS = ["en", "es"]
ROOT = Path(__file__).resolve().parent.parent
API_URL = os.environ.get('DEEPL_API_URL', 'https://api-free.deepl.com/v2/translate')

def read_api_key():
    key = os.environ.get('DEEPL_AUTH_KEY') or os.environ.get('DEEPL_API_KEY')
    key_path = ROOT / '.deepl_key'
    if not key and key_path.exists():
        key = key_path.read_text(encoding='utf-8').strip()
    if not key:
        return None
    return key

def split_front_matter(content):
    if not content.startswith('---\n'):
        return None, content
    parts = re.split(r'^---\s*\n', content, maxsplit=2, flags=re.MULTILINE)
    if len(parts) >= 3:
        # parts[1] is yaml, parts[2] is body
        return parts[1], parts[2]
    return None, content

GRADE_LINE_RE = re.compile(r':\s*[0-9]+(?:\.[0-9]+)?')

def extract_grades_block(text):
    lines = text.splitlines(True)
    n = len(lines)
    start = None
    end = None
    for i in range(n):
        if GRADE_LINE_RE.search(lines[i]):
            if start is None:
                start = i
            end = i
        elif start is not None and end is not None and i == end + 1:
            break

    if start is None:
        return text, [], ''

    # If a markdown heading (e.g. "### Notes") appears immediately above the grades block,
    # include it in the grades chunk so that it is not sent to DeepL.
    for candidate in range(max(0, start - 2), start):
        if re.match(r'^\s*#+\s+', lines[candidate]):
            start = candidate
            break

    before = ''.join(lines[:start])
    grades = [ln.rstrip('\n') for ln in lines[start:end+1]]
    after = ''
    return before, grades, after

def deepl_translate(text, target_lang, source_lang, api_key, api_url, dry_run=False):
    if not text:
        return ''
    if dry_run:
        print(f"[dry-run] translate to {target_lang} ({len(text)} chars)")
        return ''
    data = {
        'text': text,
        'source_lang': source_lang,
        'target_lang': target_lang.upper(),
        'preserve_formatting': '1',
        'split_sentences': 'nonewlines'
    }
    encoded = parse.urlencode(data).encode('utf-8')
    req = request.Request(api_url, data=encoded, method='POST')
    req.add_header('Authorization', f'DeepL-Auth-Key {api_key}')
    req.add_header('Content-Type', 'application/x-www-form-urlencoded')
    try:
        with request.urlopen(req, timeout=30) as resp:
            body = resp.read().decode('utf-8')
            j = json.loads(body)
            texts = [t.get('text','') for t in j.get('translations', [])]
            return '\n\n'.join(texts)
    except error.HTTPError as he:
        raise RuntimeError(f"DeepL API error {he.code}: {he.read().decode('utf-8')}")
    except Exception as e:
        raise

def find_template_labels(root, lang):
    dest_dir = Path(root) / '_i18n' / lang / '_posts'
    if not dest_dir.exists():
        return None
    for p in sorted(dest_dir.glob('*.md')):
        content = p.read_text(encoding='utf-8')
        _fm, body = split_front_matter(content)
        before, grades, after = extract_grades_block(body)
        if grades:
            labels = []
            for line in grades:
                # skip heading lines and empty lines
                if re.match(r'^\s*#+\s+', line) or not line.strip():
                    continue
                if ':' not in line:
                    continue
                parts = line.split(':', 1)
                labels.append(parts[0].strip())
            if labels:
                return labels
    return None

def sanitize_label(lbl):
    s = lbl.strip()
    # remove surrounding underscores or asterisks
    s = re.sub(r'^[*_]+|[*_]+$', '', s)
    s = s.rstrip(':')
    return s.strip()

FALLBACK_LABELS = {
    'en': ['Uniqueness', 'Finesse', 'Comfort', 'Intensity', 'Overall impression'],
    'es': ['Originalidad', 'Fineza', 'Reconfortante', 'Intensidad', 'Impresión general']
}

FINAL_LABEL = {'en': '**Final evaluation**', 'es': '**Nota final**'}
HEADING = {'en': '### Evaluation', 'es': '### Evaluación'}

def build_grades_block(template_labels, source_grades_lines, lang=None):
    if not source_grades_lines:
        return ''

    # strip heading lines (e.g. "### Notes") from the source grades block before parsing
    lines = [ln for ln in source_grades_lines]
    while lines and re.match(r'^\s*#+\s+', lines[0]):
        lines.pop(0)

    # extract numbers and final fraction
    numbers = []
    final_fraction = None
    for ln in lines:
        if re.match(r'^\s*#+\s+', ln):
            continue
        if '**' in ln:
            m = re.search(r'([0-9]+(?:\.[0-9]+)?\s*/\s*25)', ln)
            if m:
                final_fraction = m.group(1).replace(' ', '')
            continue
        m = re.search(r'([0-9]+(?:\.[0-9]+)?)', ln)
        if m:
            numbers.append(m.group(1))

    labels = template_labels if isinstance(template_labels, list) else None
    # infer lang fallback
    chosen = labels or FALLBACK_LABELS.get(lang, FALLBACK_LABELS['en'])
    heading = HEADING.get(lang, HEADING['en'])
    final_label = FINAL_LABEL.get(lang, FINAL_LABEL['en'])

    out = [heading]
    for idx, lbl in enumerate(chosen[:5]):
        clean = sanitize_label(lbl)
        val = numbers[idx] if idx < len(numbers) else ''
        out.append(f"_{clean}_: {val}  ")

    if final_fraction:
        out.append('\n' + f"{final_label}: {final_fraction}")
    else:
        if all(re.match(r'^\d+(?:\.\d+)?$', n) for n in numbers if n):
            total = sum(float(n) for n in numbers if n)
            total_s = str(int(total)) if total.is_integer() else str(total)
            out.append('\n' + f"{final_label}: {total_s}/25")
        else:
            out.append('\n' + f"{final_label}:")

    return '\n'.join(out)

def parse_front_matter_yaml(fm_text):
    if not fm_text:
        return {}
    if HAS_YAML:
        try:
            return yaml.safe_load(fm_text) or {}
        except Exception:
            pass
    # fallback: crude parsing for title/tags/categories
    fm = {}
    for line in fm_text.splitlines():
        m = re.match(r'^(title):\s*(?:"([^"]+)")?', line)
        if m:
            # handled below more robustly
            pass
        # try key: value
        m2 = re.match(r'^(\w+):\s*(.*)$', line)
        if m2:
            k = m2.group(1)
            v = m2.group(2).strip()
            if v.startswith('[') and v.endswith(']'):
                items = [x.strip().strip('"') for x in v[1:-1].split(',') if x.strip()]
                fm[k] = items
            else:
                fm[k] = v.strip('"')
    return fm

def emit_front_matter(fm_dict, title_val=None):
    out_lines = ['---']
    if title_val:
        escaped = title_val.replace('"', '\\"')
        out_lines.append(f'title: "{escaped}"')
    # tags
    tags = fm_dict.pop('tags', None)
    if tags:
        if isinstance(tags, list):
            items = ', '.join(f'"{t}"' for t in tags)
            out_lines.append(f'tags: [{items}]')
        else:
            out_lines.append(f'tags: ["{tags}"]')
    cats = fm_dict.pop('categories', None)
    if cats:
        if isinstance(cats, list):
            items = ', '.join(f'"{c}"' for c in cats)
            out_lines.append(f'categories: [{items}]')
        else:
            out_lines.append(f'categories: ["{cats}"]')

    # dump remaining keys if yaml available
    if fm_dict:
        if HAS_YAML:
            dumped = yaml.safe_dump(fm_dict, sort_keys=False)
            # safe_dump may include leading '---\n', ensure no duplication
            dumped = re.sub(r'^---\n', '', dumped)
            out_lines.append(dumped.rstrip('\n'))
        else:
            for k, v in fm_dict.items():
                out_lines.append(f"{k}: {v}")

    out_lines.append('---')
    return '\n'.join(out_lines) + '\n\n'

def main():
    p = argparse.ArgumentParser()
    p.add_argument('-n', '--dry-run', action='store_true', help='Show which files would be translated')
    p.add_argument('-l', '--langs', default=','.join(DEFAULT_LANGS), help='Comma-separated target langs')
    args = p.parse_args()

    options = {
        'dry_run': args.dry_run,
        'langs': [l.strip() for l in args.langs.split(',') if l.strip()],
        'source_lang': 'FR'
    }

    api_key = read_api_key()
    if not api_key and not options['dry_run']:
        print(f"ERROR: Set DEEPL_AUTH_KEY or create {ROOT / '.deepl_key'} with your key.")
        sys.exit(1)

    source_dir = ROOT / '_i18n' / 'fr' / '_posts'
    if not source_dir.exists():
        print(f"ERROR: source directory not found: {source_dir}")
        sys.exit(1)

    translated = []
    skipped = []

    for src_path in sorted(source_dir.glob('*.md')):
        filename = src_path.name
        content = src_path.read_text(encoding='utf-8')
        fm_text, body = split_front_matter(content)
        front_matter = parse_front_matter_yaml(fm_text)

        for lang in options['langs']:
            dest_dir = ROOT / '_i18n' / lang / '_posts'
            dest_dir.mkdir(parents=True, exist_ok=True)
            dest_path = dest_dir / filename
            if dest_path.exists():
                skipped.append(str(dest_path))
                continue

            if options['dry_run']:
                translated.append(str(dest_path))
                continue

            before_text, source_grades_lines, after_text = extract_grades_block(body)

            # Translate parts
            translated_before = '' if not before_text.strip() else deepl_translate(before_text, lang, options['source_lang'], api_key, API_URL, options['dry_run'])
            translated_after = '' if not after_text.strip() else deepl_translate(after_text, lang, options['source_lang'], api_key, API_URL, options['dry_run'])

            # Build grades block
            template_labels = find_template_labels(ROOT, lang)
            grades_block = '' if not source_grades_lines else build_grades_block(template_labels, source_grades_lines, lang)

            translated_body_parts = [p for p in [translated_before.strip(), grades_block.strip(), translated_after.strip()] if p]
            translated_body = '\n\n'.join(translated_body_parts) + '\n' if translated_body_parts else ''

            # Title
            translated_title = None
            if isinstance(front_matter, dict) and 'title' in front_matter:
                translated_title = deepl_translate(str(front_matter['title']), lang, options['source_lang'], api_key, API_URL, options['dry_run']).strip()

            # Tags: handle mapping for 'Noir'
            translated_tags = None
            if isinstance(front_matter, dict) and front_matter.get('tags'):
                tags = front_matter.get('tags') if isinstance(front_matter.get('tags'), list) else [str(front_matter.get('tags'))]
                out_tags = []
                for t in tags:
                    t = str(t).strip()
                    if not t:
                        continue
                    if t.lower() == 'noir':
                        mapped = 'Negro' if lang == 'es' else ('Dark' if lang == 'en' else t)
                        out_tags.append(mapped)
                    else:
                        out_tags.append(deepl_translate(t, lang, options['source_lang'], api_key, API_URL, options['dry_run']).strip())
                translated_tags = out_tags

            # Emit front matter (don't include translation metadata)
            fm_copy = dict(front_matter) if isinstance(front_matter, dict) else {}
            if 'title' in fm_copy:
                del fm_copy['title']
            if 'tags' in fm_copy:
                del fm_copy['tags']
            if 'categories' in fm_copy:
                # keep categories in fm_copy, emit below
                pass

            fm_emit = fm_copy
            # prefer translated_tags when emitting
            if translated_tags is not None:
                fm_emit['tags'] = translated_tags
            # categories unchanged

            output = emit_front_matter(fm_emit, title_val=translated_title or front_matter.get('title') if isinstance(front_matter, dict) else None)
            output += translated_body
            dest_path.write_text(output, encoding='utf-8')
            translated.append(str(dest_path))

    print(f"Translated {len(translated)} files ({', '.join(DEFAULT_LANGS)})")
    for p in translated:
        print(f"  - {p}")
    if skipped:
        print(f"Skipped {len(skipped)} existing files")

if __name__ == '__main__':
    main()
